include/unifex/win32/low_latency_iocp_context.hpp

```diff
@@ -0,0 +1,814 @@
+        operation_base* next = nullptr;
+        operation_base* prev = nullptr;
+    };
+
+    using operation_queue = intrusive_list<operation_base, &operation_base::next, &operation_base::prev>;
+
+    struct io_operation : operation_base {
+        explicit io_operation(low_latency_iocp_context& context, handle_t fileHandle, bool skipNotificationOnSuccess) noexcept
+        : operation_base(context)
+        , fileHandle(fileHandle)
+        , skipNotificationOnSuccess(skipNotificationOnSuccess)
+        , ioState(nullptr)
+        {}
+
+        const handle_t fileHandle;
+        const bool skipNotificationOnSuccess;
+        vectored_io_state* ioState;
+
+        // Cancel outstanding I/O operations (if any)
+        void cancel_io() noexcept;
+
+        // Poll for whether or not the operation has completed.
+        // Returns 'true' if the operation is completed (in which case the operation
+        // has 'acquire' semantics), 'false' otherwise. 
+        bool is_complete() noexcept;
+
+        // Start reading the next 'buffer.size()' bytes into 'buffer'.
+        //
+        // Returns 'true' if a additional read-operations can be submitted.
+        // This will only be the case if all of the following are true:
+        // - a read of the entirety of 'buffer' was successfully started
+        // - we haven't reached the maximum number of I/O operations in
+        //   this batch.
+        //
+        // Once all I/O operations have been started, the caller should
+        // schedule a 'poll' of this operation by setting this->callback
+        // and calling context.schedule_poll(this). This will schedule the
+        // callback to be run immediately (if it completed synchronously).
+        bool start_read(span<std::byte> buffer) noexcept;
+
+        // Start writing the context of 'buffer' to fileHandle.
+        bool start_write(span<const std::byte> buffer) noexcept;
+
+        std::size_t get_result(std::error_code& ec) noexcept;
+    };
+
+    struct vectored_io_state {
+        // Parent operation.
+        //
+        // May be nullptr if this operation already completed due to a poll.
+        // If this is the case then when all pending completion notifications
+        // are received then this will just go straight back on the free-list.
+        io_operation* parent = nullptr;
+
+        // Intrusive list ptr.
+        vectored_io_state* next = nullptr;
+        vectored_io_state* prev = nullptr;
+
+        // Total number of operations started
+        std::uint8_t operationCount = 0;
+
+        // Number of operations not yet received completion-notification
+        // via the IOCP. This structure is not free to be reused until
+        // this number reaches zero.
+        std::uint8_t pendingCompletionNotifications = 0;
+
+        // Whether or not the 'parent' has already been notified of completion.
+        bool completed = false;
+
+        std::array<overlapped, max_vectored_io_size> operations;
```

Comment:
Note that you can pack these more tightly than `OVERLAPPED`, which is a Win32 extension of `IO_STATUS_BLOCK`. You only actually need `IO_STATUS_BLOCK`.

Reply:
Thanks, I had considered doing this but couldn't seem to find winsock APIs equivalent to WSASend/WSARecv that take IO_STATUS_BLOCK instead of OVERLAPPED and didn't want to have to maintain separate io-state pools for files/pipes vs sockets.

Although the socket APIs have native support for vectorised I/O so it will generally only ever use the first OVERLAPPED anyway.

Maybe this could be a union of a single `overlapped` and array of `io_status_block` and just let the I/O operation choose which one to use based on the APIs it's calling?

Reply:
That's because Winsock is actually an abomination :). Internally, Windows literally cannot do scatter-gather i/o. The kernel structures don't have the facility. So, the winsock loadable does a poor emulation by issuing an IRP per buffer.

You're much better off, in my opinion, in dispensing with winsock entirely and just use Windows RIO throughout. They're far better designed APIs (they look very like io_uring, unsurprisingly), they work in terms of registered i/o buffers and submission and completion ring buffers stored in memory shared with the kernel. Your implementation would simply memcpy user buffers into a portion of those registered i/o buffers, submit an i/o request to RIO, it'll complete via IOCP using the normal `IO_STATUS_BLOCK` if you ask it to do that.

In other words, I'm saying use RIO to do socket i/o just like you do file i/o. Then everything is using the same identical IOCP mechanism, no special casing the network i/o. It's how Windows works internally in any case, so you'll be skipping an entire layer of software.

Once you're into the RIO weeds, you may wish to pay attention to Microsoft's shiny new RIO + Compute integration. I know all the Microsoft folk are super excited about this, and if you can demonstrate that libunifex uses their shiny new toys, that's a nice number of library implementer votes in favour in committee.

Reply:
@ned14 I've started looking into converting the use of `OVERLAPPED` into `IO_STATUS_BLOCK` but have been running into the question of which function I should call instead of `CancelIoEx()` to cancel an individual pending I/O.

Can I still call `CancelIoEx()` with the `IO_STATUS_BLOCK*` cast to an `OVERLAPPED*`? Or is there some other `NtCancel...()` function that I need to call instead?

Reply:
`NtCancelIoFileEx` takes a handle and two pointers to `IO_STATUS_BLOCK`. `OVERLAPPED` is simply an `IO_STATUS_BLOCK` (two void*'s) with extra stuff after it, two extra void*.

Only Win32 ever uses the extra stuff in `OVERLAPPED`. The NT kernel exclusively works in `IO_STATUS_BLOCK`. If you therefore only use the NT kernel APIs, you gain.

There are two big advantages to using `IO_STATUS_BLOCK`:

1. They are half the size, and you generally need one per i/o, so that's a nice improvement on CPU cache loading.

2. You can place your own extra stuff after each `IO_STATUS_BLOCK` instead e.g. a per-i/o pointer. The special branch of LLFIO I did Jan 2020 I think put a pointer to the Sender and a pointer to the `llfio::io_handle` or something like that. I might be completely wrong on this, but I think `IO_STATUS_BLOCK` requires an 8 byte alignment, so on 64 bit you can legally have three-void* `IO_STATUS_BLOCK`, but on 32 bit you must pad. The current LLFIO, by the way, has two-void* sized `IO_STATUS_BLOCK` and what we do is use aligned malloc to ensure the whole block of `IO_STATUS_BLOCK` is aligned to some boundary. We can then AND round down the pointer to the head of the block, and at the very front we store a pointer to `llfio::io_handle`, and from there we know how to handle things. Packing the `IO_STATUS_BLOCK` more densely was found to be a nice source of performance gain, especially as you need lots of them.

Reply:
Thanks for the help with NtCancelIoFileEx().

I've updated the PR to now deal directly with IO_STATUS_BLOCK structures, which has halved the memory used for each I/O state and it now uses the lower-level NtReadFile/NtWriteFile/NtCancelIoFileEx/NtRemoveIoCompletion/etc. APIs.

The vectored_io_state now holds 4 pointers worth of header stuff per 'sender operation', and also contains an array of 30 IO_STATUS_BLOCK entries. So each sender can issue a batch of up to 30 buffers for a scatter/gather-like operation.

Reply:
Very nice!

My Christmas holiday side project this year is to add to LLFIO some sort of mechanism which:

1. Is Sender-Receiver, Executors and Schedulers compatible (but stops just short of those, same as `llfio::io_multiplexer`).
2. Wraps the system provided self-dynamically-scaling kernel thread pool into a low level portable API which somehow (?) will integrate with any `llfio::io_multiplexer`.
3. Can pace the work submitted to the system provided kernel thread pool according to current i/o operation latency on the relevant backing hardware devices.

All this is quite hard to achieve, which is why I can't really justify writing one on work time, because chances are good I won't succeed quickly. And besides, I think WG21 is currently rather too heavy on theoretical abstraction, and too light on real world practically useful implementations of all this stuff they propose standardising, in my opinion.

I suppose I better explain what specifically I want to solve here:

1. I want the sockets to all be using Windows RIO.
2. I want Windows RIO's IOCP interface to be wired into the Win32 i/o aware thread pool, which self adjusts its concurrency.
3. I want all kernel thread level concurrency to use the Win32 i/o aware thread pool.
4. With all those three Windows facilities wired together as the Microsoft folk designed and optimised the snot out of them, I want all that to appear as natural and portable low level APIs in standard C++.
5. And for that natural and portable low level API to also "do the right thing" with Grand Central Dispatch on Mac OS and other kernel thread concurrency + i/o platform-specific integrations. You might notice here that if the API is designed right you get "for free" hardware accelerated TLS sockets, or any other kind of socket-like abstraction combined with concurrency framework which your system can provide.
6. And finally, I want the socket i/o concurrency solution to work well (on Linux and Windows at least) with memory mapped file i/o.

That final point will need a bit of explaining. The problem needing solved is that the system provided dynamically self adjusting thread pool generally tries to keep as many threads active as there is hardware concurrency. So, if a thread stalls on a page fault or syscall or i/o, more kernel threads are created, whereas if more kernel threads are running non-blocked than hardware concurrency, then kernel threads are removed. The problem I need to solve is that if you point such a pool at lots of memory mapped file i/o, and you're not doing a ton of CPU work, as storage becomes congested, the thread pool "runs away" with adding lots more concurrency. This makes the problem exponentially worse until the system appears to hang because you overload storage.

So what I basically need is some sort of Kalman filter on total process concurrency whereby we cut back concurrency on non-socket i/o if that i/o is congested, but can still increase concurrency for non-i/o bound work and socket i/o, because the system thread pool has special support for being aware of socket i/o.

Anyway all the above turned into rather longer than I intended, but perhaps now you can understand the difficulty of what I'll be attempting, but also why if I'm successful, how extremely useful it would be. At work in our custom DB, we currently hand tune the thread counts which is brittle and prone to pathological corner case performance drop outs. If we could replace that with what I have in mind, the system would self-optimise itself, and even better, would self adjust over time to remain optimal.

Anyway merry christmas to all of you, and have a great holiday!